File: packages/mobile/lib/controllers/game-play.controller.dart
import 'dart:convert';

import 'package:http_interceptor/http/http.dart';
import 'package:mobile/classes/actions/action-data.dart';
import 'package:mobile/classes/game/game-message.dart';
import 'package:mobile/classes/game/game-update.dart';
import 'package:mobile/constants/endpoint.constants.dart';
import 'package:mobile/constants/socket-events/game-events.dart';
import 'package:mobile/locator.dart';
import 'package:mobile/services/socket.service.dart';
import 'package:rxdart/rxdart.dart';

import '../services/client.dart';

class GamePlayController {
  GamePlayController._privateConstructor() {
    configureSocket();
  }

  static final GamePlayController _instance =
      GamePlayController._privateConstructor();

  factory GamePlayController() {
    return _instance;
  }

  final httpClient = getIt.get<PersonnalHttpClient>();
  SocketService socketService = getIt.get<SocketService>();

  final String baseEndpoint = GAME_ENDPOINT;
  InterceptedHttp get http => httpClient.http;

  String? currentGameId;

  Stream<void> get actionDoneEvent => _actionDone$.stream;

  Stream<GameUpdateData> get gameUpdateEvent => gameUpdate$.stream;

  Stream<GameMessage?> get messageEvent => gameMessage$.stream;

  final BehaviorSubject<GameUpdateData> gameUpdate$ =
      BehaviorSubject<GameUpdateData>();
  final BehaviorSubject<GameMessage?> gameMessage$ =
      BehaviorSubject<GameMessage?>.seeded(null);
  final PublishSubject<void> _actionDone$ = PublishSubject<void>();

  Future<void> sendAction(ActionData actionData) async {
    Uri endpoint = Uri.parse("$baseEndpoint/$currentGameId/action");
    http
        .post(endpoint, body: jsonEncode(actionData))
        .then((_) => _actionDone$.add(null));
  }

  Future<void> leaveGame() async {
    Uri endpoint = Uri.parse("$baseEndpoint/$currentGameId/leave");
    await http.delete(endpoint);
  }

  void configureSocket() {
    socketService.on(GAME_UPDATE_EVENT_NAME, (dynamic newData) {
      gameUpdate$.add(GameUpdateData.fromJson(newData));
    });
    socketService.on(GAME_MESSAGE_EVENT_NAME, (dynamic gameMessage) {
      gameMessage$.add(GameMessage.fromJson(gameMessage));
    });
  }
}

File: packages/client/src/app/controllers/game-play-controller/game-play.controller.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { ActionData } from '@app/classes/actions/action-data';
import { Position } from '@app/classes/board-navigator/position';
import GameUpdateData from '@app/classes/communication/game-update-data';
import { Message } from '@app/classes/communication/message';
import { HTTP_ABORT_ERROR } from '@app/constants/controllers-errors';
import SocketService from '@app/services/socket-service/socket.service';
import { BehaviorSubject, Observable, Subject } from 'rxjs';
import { environment } from 'src/environments/environment';

@Injectable({
    providedIn: 'root',
})
export class GamePlayController {
    private gameUpdate$ = new BehaviorSubject<GameUpdateData>({});
    private newMessage$ = new BehaviorSubject<Message | null>(null);
    private firstSquareSelectedPosition$ = new BehaviorSubject<Position | null>(null);
    private actionDone$ = new Subject<void>();

    constructor(private http: HttpClient, private readonly socketService: SocketService) {
        this.configureSocket();
    }

    sendAction(gameId: string, action: ActionData): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/players/action`;
        this.http.post(endpoint, action).subscribe(() => {
            this.actionDone$.next();
        });
    }

    sendMessage(gameId: string, message: Message): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/players/message`;
        this.http.post(endpoint, message).subscribe();
    }

    sendError(gameId: string, message: Message): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/players/error`;
        this.http.post(endpoint, message).subscribe();
    }

    handleReconnection(gameId: string, newPlayerId: string): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/players/reconnect`;
        this.http.post(endpoint, { newPlayerId }).subscribe();
    }

    handleDisconnection(gameId: string): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/players/disconnect`;
        // When reloading the page, a disconnect http request is fired on destruction of the game-page component.
        // In the initialization of the game-page component, a reconnect request is made which does not allow the
        // server to send a response, triggering a Abort 0  error code which is why we catch it if it this this code
        this.http.delete(endpoint, { observe: 'response' }).subscribe(this.handleDisconnectResponse, this.handleDisconnectError);
    }

    handleFirstSquareSelected(gameId: string, position: Position): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/squares/select`;
        this.http.post(endpoint, position).subscribe();
    }

    handleFirstSquareCancelled(gameId: string): void {
        const endpoint = `${environment.serverUrl}/games/${gameId}/squares/cancel`;
        this.http.delete(endpoint).subscribe();
    }

    observeGameUpdate(): Observable<GameUpdateData> {
        return this.gameUpdate$.asObservable();
    }

    observeNewMessage(): Observable<Message | null> {
        return this.newMessage$.asObservable();
    }

    observeActionDone(): Observable<void> {
        return this.actionDone$.asObservable();
    }

    observeFirstSquareSelectedPosition(): Observable<Position | null> {
        return this.firstSquareSelectedPosition$.asObservable();
    }

    private configureSocket(): void {
        this.socketService.on('gameUpdate', (newData: GameUpdateData) => {
            this.gameUpdate$.next(newData);
        });
        this.socketService.on('newMessage', (newMessage: Message) => {
            this.newMessage$.next(newMessage);
        });
        this.socketService.on('firstSquareSelected', (position: Position) => {
            this.firstSquareSelectedPosition$.next(position);
        });
        this.socketService.on('firstSquareCancelled', () => {
            this.firstSquareSelectedPosition$.next(null);
        });
    }

    private handleDisconnectResponse(): void {
        return;
    }

    private handleDisconnectError(error: { message: string; status: number }): void {
        if (error.status !== HTTP_ABORT_ERROR) throw new Error(error.message);
    }
}
